initial commit of Copy-SqlSsisCatalog

schwartzmx opened the pull request:
- Completely open to changing the cmdlet name
- This is an initial working function,  looking for some initial review/pointers/testing and potentially some ideas to maybe better approach this.

Cheers

---
**To Do**
- [x] rename module to be catalog specific
- [x] prompt user for password to create catalog if it doesn't exist.  Use Read-Host as secure string x 2 and check for matching to be safe

Edit: This will require loading another assembly
```powershell
$null = [System.Reflection.Assembly]::LoadWithPartialName("Microsoft.SqlServer.Management.IntegrationServices")
```

ctrlbold commented on functions/Copy-SqlIntegrationServices.ps1:
```diff
@@ -0,0 +1,392 @@
+            Write-Error $_ 
+        }
+
+        $sourceCatalog = $sourceSSIS.Catalogs | ? { $_.Name -eq "SSISDB" }  
```
Please only use full names so replace ? with Where-Object

schwartzmx commented:
Thanks @ctrlbold, reverted the psd1 and psm1 files

ctrlbold commented on functions/Copy-SqlIntegrationServices.ps1:
```diff
@@ -0,0 +1,392 @@
+                [String]$Environment,
+                [Switch]$Force
+            )
+            $f = $destinationFolders | ? { $_.Name -eq $Folder }
```
I skipped this part and later on, it was not immediately obvious what $f is. Please make this variable name more descriptive.

ctrlbold commented on functions/Copy-SqlIntegrationServices.ps1:
```diff
@@ -0,0 +1,392 @@
+    PROCESS
+    {
+        if (!$sourceCatalog) {
+            Write-Error "The source SSISDB catalog does not exist."
```
Let's talk about a future Dynamic Parameter that auto populates your $sourceCatalogs (if its possible/easy?), then you won't have to check and people can tab through :)

schwartzmx replied:
not sure if it's worth it.  as far as I know, SSISDB is the only available catalog name

ctrlbold commented on functions/Copy-SqlIntegrationServices.ps1:
```diff
@@ -0,0 +1,392 @@
+                    }
+                }
+                else {
+                    If ($Pscmdlet.ShouldProcess($Destination, "Creating folder $folder")) {
```
hell yeah!

ctrlbold commented on functions/Copy-SqlIntegrationServices.ps1:
```diff
@@ -0,0 +1,392 @@
+        }
+        else {
+            foreach ($srcFolder in $sourceFolders) {
+                if($($destinationFolders.Name) -notcontains $($srcFolder.Name)) {  
```
You don't have to $() here - plain ol' $destinationFolders.Name would work. I outlined below where the $() is needed.

schwartzmx replied:
I tend to overuse this, I think cause it's bitten me so much in the past :)
I'll fix

ctrlbold commented on functions/Copy-SqlIntegrationServices.ps1:
```diff
@@ -0,0 +1,392 @@
+        else {
+            foreach ($srcFolder in $sourceFolders) {
+                if($($destinationFolders.Name) -notcontains $($srcFolder.Name)) {  
+                    If ($Pscmdlet.ShouldProcess($Destination, "Creating folder $($srcFolder.Name)")) {
```
Here, you do have to have use the $() wrap because it's embedded in a string

ctrlbold commented on functions/Copy-SqlIntegrationServices.ps1:
```diff
@@ -0,0 +1,392 @@
+                        Write-Warning "Integration services catalog folder $folder exists at destination. Use -Force to drop and recreate."
+                    }
+                    else {
+                        If ($Pscmdlet.ShouldProcess($Destination, "Dropping folder $folder and recreating")) {
```
Love this support for WhatIf 👍

schwartzmx commented:
thanks for the review @ctrlbold,  I've attended to your comments/suggestions.  I also renamed the function, and added functionality for prompting and creating the destination catalog if it doesn't exist.

Thanks.   Have a good weekend 👍

ctrlbold commented on functions/Copy-SqlSsisCatalog.ps1:
```diff
@@ -0,0 +1,420 @@
+    {
+        if (!$sourceCatalog) {
+            Write-Error "The source SSISDB catalog does not exist."
+            exit
```
Sorry, a couple more that I missed from yesterday. Write-Error is non terminating errors and throw is for terminating errors. In this case, you'd want to  

throw "The source SSISDB catalog does not exist."

Also, please change all exits to returns or throws. Exit can close a PowerShell window and is a bit too harsh. Return is a better alternative.

schwartzmx replied:
No problem.  Thanks for the suggestions.

In my last commit I've altered the `exit`s and updated the `write-error`s to `throw` in the appropriate places for terminating errors.

Thanks 👍

dansqldba replied:
@schwartzmx have you incorporated all of the points from the review(s) provided by @ctrlbold? Is this cmdlet ready for QA?

schwartzmx commented:
@dansqldba I think this is good for QA now 👍

ctrlbold commented:
Added required assembly

dansqldba commented:
- [ ] Working/useful help content, including link to command on dbatools web site
- [ ] All examples work as advertised
- [ ] Does not contain template content
- [ ] Does not contain excessive/unnecessary amounts of comments
- [ ] Works remotely
- [ ] Works locally
- [ ] Works on lower versions or throws error specifying version not supported
- [ ] Works with named instances
- [ ] Works with clustered instances
- [ ] Handles offline/read only databases
- [ ] Supports multiple servers (at the command line or piped from Get-SqlRegisteredServerName)
- [ ] No un-handled errors which stop the command working with multiple servers

ctrlbold commented on functions/Copy-SqlSsisCatalog.ps1:
```diff
@@ -0,0 +1,472 @@
+This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+You should have received a copy of the GNU General Public License along with this program.  If not, see <http://www.gnu.org/licenses/>.
+.LINK
+https://dbatools.io/functions/Copy-SqlSsisCatalog
```
The shortlink is valid -- this hsould be https://dbatools.io/Copy-SqlSsisCatalog

ctrlbold commented on functions/Copy-SqlSsisCatalog.ps1:
```diff
@@ -0,0 +1,472 @@
+            }
+
+            $catalog = New-Object "$ISNamespace.Catalog" ($destinationSSIS, "SSISDB", $plainTextPass)  
+            $catalog.Create()
```
Interesting - you must first change to enable clr if it's not enabled. Please prompt to confirm.
Exception calling "Create" with "0" argument(s): "Integration Services uses CLR stored procedures. Enable "clr enabled" configuration option."
At C:\github\branches\dbatools\functions\Copy-SqlSsisCatalog.ps1:207 char:13
+             $catalog.Create()
+             ~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : IntegrationServicesException

ctrlbold commented on functions/Copy-SqlSsisCatalog.ps1:
```diff
@@ -0,0 +1,472 @@
+                    else {
+                        if ($Pscmdlet.ShouldProcess($Destination, "Dropping folder $folder and recreating")) {
+                            try {
+                                New-CatalogFolder -Folder $srcFolder.Name -Description $srcFolder.Description -Force
```
One of these commands is causing issues, i think it's this one

PS C:\github\branches> Copy-SqlSsisCatalog -Source sql2014 -Destination sql2016
WARNING: Integration services catalog folder foo exists at destination. Use -Force to drop and recreate.
Integration services migration finished.
PS C:\github\branches> Copy-SqlSsisCatalog -Source sql2014 -Destination sql2016 -Folder
Copy-SqlSsisCatalog : Missing an argument for parameter 'Folder'. Specify a parameter of type 'System.String' and try again.
At line:1 char:58
+ Copy-SqlSsisCatalog -Source sql2014 -Destination sql2016 -Folder

schwartzmx replied:
@ctrlbold is this part the issue?
>Copy-SqlSsisCatalog -Source sql2014 -Destination sql2016 -Folder
Looks like Folder wasn't provided, may have meant `Force`?

ctrlbold commented on functions/Copy-SqlSsisCatalog.ps1:
```diff
@@ -0,0 +1,472 @@
+        }
+        catch {
+            Write-Exception $_
+            throw "An error occured when checking the destination for Integration Services."
```
I get this when IS isn't installed. Please add 

throw "An error occured when checking the destination for Integration Services. Is Integration Services installed?"

schwartzmx commented:
@ctrlbold re: your comment on splitting these up into smaller more granular functions.

I think that's a good idea, I'd be able to do that in the future.   Maybe we want to create an issue and assign it to me?

----------

also I've attended to the comments, if you have a chance to retest.   Also added the ability to skip the prompt with a switch param, `EnableSqlClr` for more automated way of calling this function (same as what we did with the Create catalog password as a secure string param).
